flocker/node/agents/functional/test_cinder.py

```diff
@@ -802,79 +802,93 @@ def test_no_immediate_authentication(self):
+        backend, api_args = backend_and_api_args_from_configuration({
+            "backend": "openstack",
+            "auth_plugin": "rackspace",
+            "region": "ORD",
+            "username": "mimic",
+            "api_key": "12345",
+            "auth_url": "http://127.0.0.1:{}/identity/v2.0".format(
+                listening_port.getHost().port
+            ),
+        })
+        # Cause the Mimic server to close incoming connections
+        patch.patch()
+        api = get_api(
+            backend=backend,
+            api_args=api_args,
+            reactor=object(),
+            cluster_id=make_cluster_id(TestTypes.FUNCTIONAL),
+        )
+        # List volumes with API patched to close incoming connections.
+        try:
+            result = api.list_volumes()
+        except ConnectFailure:
+            # Can't use self.asserRaises here because that would call the
+            # function in the main thread.
+            pass
+        else:
+            self.fail(
+                'ConnectFailure was not raised. '
+                'Got {!r} instead.'.format(
+                    result
+                )
             )
+        finally:
+            # Re-enable the Mimic server.
+            # The API operations that follow should succeed.
             patch.restore()
-            return api
-
-        mimic_started = mimic_starting.addCallback(build_api)
-
-        def list_volumes(api, force_connection_failure=False):
-            if force_connection_failure:
-                patch.patch()
-            try:
-                return api.list_volumes()
-            finally:
-                patch.restore()
-
-        def check_failing_connection(api):
-            d = deferToThread(
-                lambda api: list_volumes(api, force_connection_failure=True),
-                api,
-            )
-            d = self.assertFailure(d, ConnectFailure)
-            # return the api for further testing.
-            d = d.addCallback(
-                lambda failure_instance: api
-            )
-            return d
-        listing_volumes1 = mimic_started.addCallback(check_failing_connection)
 
-        def check_successful_connection(api):
-            d = deferToThread(
-                lambda api: list_volumes(api, force_connection_failure=False),
-                api,
-            )
-            d = d.addCallback(
-                lambda result: self.assertEqual([], result)
-            )
-            return d
-        finishing = listing_volumes1.addCallback(check_successful_connection)
+        # List volumes with API re-enabled
+        result = api.list_volumes()
+        self.assertEqual([], result)
+
+        # Close the connection from the client side so that the mimic server
+        # can close down without leaving behind lingering persistent HTTP
+        # channels which cause dirty reactor errors.
+        # XXX: This is gross. Perhaps we need ``IBlockDeviceAPI.close``
+        (api
```

Comment:
this gets the client session and closes, yes agreed this is ugly, but what benefits would IBlockDeviceAPI.close() have other than code clarity? (assessing whether is worth adding instead of having this here)